### ISSUE TITLE
a demo that civetweb support mbedtls

### ISSUE THREAD

**Issue by Qinch** [civetweb/civetweb#839]
here is a demo that civeweb v1.11 supports mbedtls v2.16.2

https://github.com/Qinch/civetweb/tree/v1.11-mbedtls

**Comment by bel2125** [comment]
I had a look at your implementation.
It has been done on an older version of CivetWeb, so it has a couple of merge conflicts.
But the changes to "civetweb.c" are not too much, so I think they can be transferred easily to the current version.
Probably it could be added as a compile option to use MbedTLS instead of OpenSSL?
But I don't want to add the MbedTLS sources directly into the civetweb source tree.

**Comment by Qinch** [comment]
HI @bel2125 
Thanks for your reply. 
I use civetweb in an arm environment, so implement a demo which using MbedTLS( the compile options is `-DNO_SSL -DMG_MBEDTLS`). I think it could be added as a compile option 
I'm sorry for my broken english.

**Comment by Qinch** [comment]
@bel2125  I think it could be added as a compile option to use MbedTLS instead of OpenSSL.
I wold like to submit a merge request to the recent version. 
By the way, I'm sorry for my broken english.

**Comment by bel2125** [comment]
Don't worry about your English ;-)

Yes, I think it would be valuable to have an alternative to OpenSSL for CivetWeb. The Apache licensed version of mbetTLS seems to be a good candidate.
If you provide a pull request for the current master, I will be happy to merge it - but give me a few days to review it.

**Comment by Qinch** [comment]
OK, give me some days. I will provide a pull request.
In addition, civetweb helps me a lot. Thanks.

**Comment by laoshaw** [comment]
is this merged? interested in testing, always nice to have alternatives for openssl and it will be easier to build civetweb on windows/macos with mbedtls comparing to openssl too.

**Comment by bel2125** [comment]
I am afraid it is not yet merged - I did not get a pull request.
Probably the current HEAD version uses a couple of additional TLS functions, not used in V1.11.
I still think it would be valuable to merge it.
@Qinch : If I create a pull request from your branch, could we go through it together, probably using some GitHub comments on some code sections.

**Comment by Qinch** [comment]
@bel2125 I'm really busy recently. I've been off work at 11 o'clock every night recently.
I will submit an MergeRequest as soon as possible.

**Comment by Qinch** [comment]
@bel2125 I am very sorry. I've been too busy with my work recently。
I will try to do it this Sunday.
```bash

by the way , is it necessary to rework the TLS code,  like this:
                         civetweb    
                              |
                              |
                abstract tls functions
                /                    \
              /                        \
   openssl functions          mbedtls functions

```

**Comment by bel2125** [comment]
I hope I did not ruin your weekend ... just take the time you need.
If you think, it is better to introduce some abstract tls functions first, we can do so as well. You definitely know more on mbedTLS than I do.

Civetweb currently uses 66 OpenSSL API functions (42 from libssl, 24 from libcrypto), so probably this is not a 1:1 replacement anyway.

**Comment by laoshaw** [comment]
openwrt has libustream as a thin layer to wrap various ssl libraries, as a matter of fact it is switching from mbedtls to wolfssl as the default ssl library, fyi

**Comment by bel2125** [comment]
@laoshaw Thank you for this information. ustream-ssl seems to be able to switch between openssl, mbedtls and wolfssl. 
So supporting libustream might allow to not only choose mbedtls but also wolfssl (although wolfssl also has an openssl compatibility layer).
Do you know where the official repository is? I found this page: https://lxr.openwrt.org/source/ustream-ssl/

**Comment by laoshaw** [comment]
https://git.openwrt.org/?p=project/ustream-ssl.git;a=summary

**Comment by bel2125** [comment]
Thank you very much.
I was busy finalizing and testing V1.13 release the last week, so I did not have the time to take a deeper look - I will do this next week.

It seems ustream provides just the fundamental APIs: read, write, accept, ... but not the more advanced (or rarely used?) features such as client certificate checks or ALPN. But I will definitely need some more time to check this in detail.

**Comment by xhpohanka** [comment]
Were you able to upload a (bigger) file in civetweb POST using TLS over mbedtls? I have tested it only on zephyr rtos but it always fails.

**Comment by bel2125** [comment]
@xhpohanka : I don't know what "bigger" means in bytes.

The current mbedtls branch is still in a test state - in particular, it debug-prints all read/written data to the console, and this does slow down all tests, of course. To test with bigger data, remove the fprintf(stdout, "mbedtls: mbedtls_ssl_read/write %d\n", rc);` in mod_mbedtls.inl

I just tested on Linux (x64 Ubuntu in WSL). I followed the install description in doc/mbedtls.md, except I wanted to add WITH_LUA=1 and WITH_WEBSOCKET=1 for my test, and needed to run "ldconfig" once before starting the server. 
Indeed, it seems I can upload files up to ~13 kB without problems, while file with ~16 kB or more cause an error in my form POST handler. Also, the "1000images" test misses a lot of images when loading, and after this test no other page can be loaded anymore. 
It is still in a test phase.

**Comment by bel2125** [comment]
I guess there is an issue with closing the connections, since I can only do "num_threads" requests at most. By turning "num_thread" to one, the second reload of a page already returns an error, until I restart the server.

**Comment by xhpohanka** [comment]
The Zephyr has its own secure socket implementation built around mbedtls (see https://docs.zephyrproject.org/2.3.0/reference/networking/sockets.html#secure-sockets). I have tried to alter civetweb's source to use this secure sockes feature instead of internal TLS support. The result is that it behaves the same. Page loading is working correctly but file upload is still broken. I cannot find if the issue is in mbedtls or civetweb itself...

**Comment by bel2125** [comment]
Probably there is some issue with the integration of the `mbedtls_ssl_read` function:
https://github.com/civetweb/civetweb/blob/f6c64cd4742cc7832bbca1ff2cf0b7d3e1c1efc7/src/mod_mbedtls.inl#L161
Maybe it needs to be called in a loop, maybe the return value needs to be evaluated differently ...
I won't have time this week to look at this in more detail, but I would start looking somewhere around there (probably I'm wrong).

**Comment by Qinch** [comment]
@bel2125  Hi
How to execute the test case? I would like to look at this issue this Sunday.

**Comment by xhpohanka** [comment]
> How to execute the test case? I would like to look at this issue this Sunday.

Unfortunately I have no time to prepare simple test application. But if you are able to build a small site with single form and file input, then it is a test case. Uploading of files larger that several kilobytes fails on my system.

**Comment by bel2125** [comment]
I am testing using Ubuntu in WSL.

Terminal 1 (server):
```
cd /home/of/civetweb/
git checkout mbedtls
make clean
make build WITH_LUA=1 COPT="-DNO_SSL -DUSE_MBEDTLS" LOPT="-lmbedcrypto -lmbedtls -lmbedx509"
sudo setcap CAP_NET_BIND_SERVICE=+eip civetweb
./civetweb -document_root test -num_threads 5 -listening_ports 80,443s -ssl_certificate resources/cert/server.pem
```

I'm building and starting the server using the default http and https ports (and set port bind permissions using setcap, but you can also use sudo or work as root ... does not matter for this test). Note that I limited the number of threads to only 5. Then I used a browser to open http://localhost and hit "refresh" for about 10 times - and it always works.
Now if I do the same test for https://localhost, it only works 5 times - if I set the number or threads to 2, it only works 2 times (probably you need to accept the self signed certificate once, so the first test might be not meaningful).

Then I do tests with POST requests using `curl`. First I create a random binary file with 256k, then send it using http and https.
The behavior is similar to the browser: Only the first "num_thread" calls work. In addition, there seems to be a size limit.

Terminal 2 (test client):
```
dd if=/dev/random of=testfile.bin bs=256 count=1024

curl --header "Content-Type:application/octet-stream" --data-binary @testfile.bin -H 'Expect:' --output testfile.ret http://localhost/page6.lua

curl --header "Content-Type:application/octet-stream" --data-binary @testfile.bin -H 'Expect:' --output testfile.ret -k https://localhost/page6.lua
```

**Comment by Rodolphe06** [comment]
Hello, 

I am porting CivetWeb from Openssl to mbedtls.
I have a dedicated Client which works well with Civetweb and openssl.
Concerning mbedtls, Everything works except as you found out the issue concerning the "num_threads".

@bel2125 
Could you please give me a hint on how to find the issue ?
Why would mbedtls need several thread to have a continuous valid tls communication ?

When using netstat, i can see that sockets are well closed after a certain time so it seems not the issue.
master thread is in poll and worker threads are waiting for data in consume_socket.
 
Only weird fact is that when it is not working with a small amount of num_threads, the https request contains only "GET" and therefor we face Error 400 Malformed Message

**Comment by bel2125** [comment]
> Why would mbedtls need several thread to have a continuous valid tls communication ?

I think this is simply a bug that still exists in this alpha version of the mbedtls port - this is still in a feature development branch and not in the master branch. Finally it should not need any more threads as compared to OpenSSL. 
I did not have a look since the end of November - probably it is just a small bug with a big effect.
I will post an update once we know more.

**Comment by Qinch** [comment]
@bel2125 @Rodolphe06 
hello
when num_threads is 1, civetweb(the mbedtls branch) is not working well . but the v1.11 version seems to works OK.
the url is :https://116.85.50.74:2221/

the operations as follow:

```bash

https://github.com/Qinch/civetweb/tree/v1.11-mbedtls


root@10-255-1-58:~/v11/civetweb# ./civetweb -listening_ports 888,2221s -ssl_certificate resources/cert/server.pem -document_root ./test/htmldir/ -num_threads 1 &


```

**Comment by Rodolphe06** [comment]
Hello @bel2125 and @Qinch ,

Thanks a lot for the info. I will check that.

BR,

Rodolphe

**Comment by bel2125** [comment]
I did compare your work based on the V1.11 with the unmodified V1.11 tag:
https://github.com/civetweb/civetweb/compare/v1.11...Qinch:v1.11-mbedtls
in particular src/mod_mbedtls.inl and src/civetweb.c:

There seem to be no substantial difference in the mod_mbedtls.ini from V1.11 and the current one, basically some symbol names and comments, plus changes in mbed_ssl_handshake, but handshake does not seem to be the problem now.

Of course there are many changes to civetweb.c since V1.11 - so I looked at selected differences in the v1.11-mbedtls version from v1.11 and checked if they are still present in the head version. It seems the MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS checks are only in the head version. One interesting difference: the v1.11-mbedtls used to comment the "set_non_blocking_mode" call:

![image](https://user-images.githubusercontent.com/5619191/102765421-e2805900-437c-11eb-9081-5dc65c524cad.png)

**Comment by Rodolphe06** [comment]
Hi @Qinch 

As you said, i confirm v1.11 is working fine as expected.

Concerning the set_non_blocking_mode, it does not correct the issue on mbed branch source code.

**Comment by Qinch** [comment]
!defined NO_SSL  macro  may  be need to be ported.  such as  ! defined NO_SSL or  defined use_mbedtls. 
I will try to look at this problem.

**Comment by xhpohanka** [comment]
Hello,

I have found the reason of uploading issues using this mbedtls implementation.
It is caused by the fact, that mbedtls code does not implement reading data remaining in tls layer same as for the case of openssl and polls unconditionally
https://github.com/civetweb/civetweb/blob/935b493e0c60795ef1d8a67a0f5f10c3a47d3db2/src/civetweb.c#L6767

We need there something similar, eg...

```
to_read = mbedtls_ssl_get_bytes_avail(conn->ssl);

if (to_read > 0) {
    /* We already know there is no more data buffered in conn->buf
     * but there is more available in the SSL layer. So don't poll
     * conn->client.sock yet. */

    pollres = 1;
    if (to_read > len)
        to_read = len;

    DEBUG_TRACE("mbedtls read remaining %d bytes", to_read);
}
else {
	 //DO POLLING
}
...
```

I have not finished my testing yet, but it seems that it solves my issues now...

**Comment by xhpohanka** [comment]
There is still an issue when `MBEDTLS_SSL_MAX_CONTENT_LEN` is defined and smaller that 16kB, which is the case of a zephyr's  in-tree mbedtls.
In such case there are mbedssl reads returning `-0x7200`...

```
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:8108 str:<= handshake
mbedtls handshake rc:0 state:16
mbedtls mbed_ssl_accept state:16
*** 1610548143.764289990 140310536840768 process_new_connection:18592: Start processing connection from 127.0.0.1
*** 1610548143.764296689 140310536840768 process_new_connection:18599: calling get_request (1 times for this connection)
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:8284 str:=> read
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:4325 str:=> read record
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2534 str:=> fetch input
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2694 str:in_left: 0, nb_want: 5
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2718 str:in_left: 0, nb_want: 5
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2720 str:ssl->f_recv(_timeout)() returned 5 (-0xfffffffb)
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2740 str:<= fetch input
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2534 str:=> fetch input
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2694 str:in_left: 5, nb_want: 261
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2718 str:in_left: 5, nb_want: 261
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2720 str:ssl->f_recv(_timeout)() returned 256 (-0xffffff00)
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2740 str:<= fetch input
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:1792 str:=> decrypt buf
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2374 str:<= decrypt buf
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:4399 str:<= read record
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:8572 str:<= read
*** 1610548143.764585893 140310536840768 switch_domain_context:13903: HTTPS Host: localhost
*** 1610548143.764600793 140310536840768 process_new_connection:18685: http: 1.1, error: none
*** 1610548143.764606668 140310536840768 handle_request:14457: URL: /bigfile
*** 1610548143.764625269 140310536840768 open_auth_file:8842: fopen(/.htpasswd): No such file or directory
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:8284 str:=> read
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:4325 str:=> read record
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2534 str:=> fetch input
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2694 str:in_left: 0, nb_want: 5
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2718 str:in_left: 0, nb_want: 5
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2720 str:ssl->f_recv(_timeout)() returned 5 (-0xfffffffb)
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2740 str:<= fetch input
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:4108 str:bad message length
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:4358 str:ssl_get_next_record() returned -29184 (-0x7200)
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:8349 str:mbedtls_ssl_read_record() returned -29184 (-0x7200)
SSL read failed, error -29184
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:8696 str:=> write
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:3351 str:=> write record
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:1443 str:=> encrypt buf
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:1779 str:<= encrypt buf
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2753 str:=> flush output
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2771 str:message length: 133, out_left: 133
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2777 str:ssl->f_send() returned 133 (-0xffffff7b)
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2805 str:<= flush output
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:3484 str:<= write record
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:8724 str:<= write
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:8696 str:=> write
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:3351 str:=> write record
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:1443 str:=> encrypt buf
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:1779 str:<= encrypt buf
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2753 str:=> flush output
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2771 str:message length: 85, out_left: 85
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2777 str:ssl->f_send() returned 85 (-0xffffffab)
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2805 str:<= flush output
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:3484 str:<= write record
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:8724 str:<= write
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:8284 str:=> read
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:4325 str:=> read record
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2534 str:=> fetch input
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2694 str:in_left: 5, nb_want: 5
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2740 str:<= fetch input
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:4108 str:bad message length
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:4358 str:ssl_get_next_record() returned -29184 (-0x7200)
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:8349 str:mbedtls_ssl_read_record() returned -29184 (-0x7200)
SSL read failed, error -29184
*** 1610548143.764852432 140310536840768 process_new_connection:18707: handle_request done
*** 1610548143.764857727 140310536840768 process_new_connection:18779: Done processing connection from 127.0.0.1 (0.000000 sec)
mbedtls close
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:8739 str:=> write close notify
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:5264 str:=> send alert message
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:3351 str:=> write record
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:1443 str:=> encrypt buf
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:1779 str:<= encrypt buf
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2753 str:=> flush output
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2771 str:message length: 69, out_left: 69
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2777 str:ssl->f_send() returned 69 (-0xffffffbb)
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:2805 str:<= flush output
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:3484 str:<= write record
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:5277 str:<= send alert message
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:8755 str:<= write close notify
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:8948 str:=> free
file:/home/honza/dev/armmbed/mbedtls/library/ssl_tls.c line:9013 str:<= free
*** 1610548143.764978449 140310536840768 worker_thread_run:19116: Connection closed
*** 1610548143.764983380 140310536840768 consume_socket:18864: going idle
```

**Comment by bel2125** [comment]
I think I found the reason why I cannot run more than `num_thread` requests:
In `worker_thread_run` the `init_connection` call is skipped for the mbedtls case.
Since V1.11 this piece of code was moved here from another function during some restructuring (while implementing buffered response headers and HTTP2). I will push a fix, but I need to remove some instrumentation before.

@xhpohanka: 
The 16kB limit of `MBEDTLS_SSL_MAX_CONTENT_LEN` seems to be normal for "general purpuse use" of TLS: https://docs.zephyrproject.org/1.13.0/reference/kconfig/CONFIG_MBEDTLS_SSL_MAX_CONTENT_LEN.html
This 16 kB is a "standard server limit" (https://tls.mbed.org/discussions/generic/help-with-small-mbedtls_ssl_max_content_len), there is a way to negotiate a smaller limit down to 4 kB (https://tools.ietf.org/html/rfc6066#section-4), but this must be requested by the client using a TLS extension field.

**Comment by bel2125** [comment]
After several hours of manual merging, all conflicts seem to be resolved and the result seems to work with OpenSSL and mbedTLS:
https://github.com/civetweb/civetweb/pull/965
The code still needs some refactoring, e.g., there should be no `#define NO_SSL` if mbedTLS is used, but probably something like `#define MBEDTLS_API` (in analogy to `#define OPENSSL_API_*`), and some features available with OpenSSL are not implemented yet. But this can be done later, step by step.

**Comment by bel2125** [comment]
The mbedTLS branch is merged into master and part of the release of V1.14.
It does not support all "advanced" features we have with OpenSSL (e.g,, ALPN/HTTP2 is missing), but it contains all "standard" features required for HTTPS connections. Thus, I think we can close this issue.

Thank you everyone for your contributions.
